File: src/main/java/com/github/jaksonlin/pitestintellij/inspectors/UnittestInspector.java
package com.github.jaksonlin.pitestintellij.inspectors;

import com.github.jaksonlin.pitestintellij.MyBundle;
import com.github.jaksonlin.pitestintellij.commands.unittestannotations.UnittestFileInspectorCommand;
import com.github.jaksonlin.pitestintellij.context.CaseCheckContext;
import com.intellij.codeInspection.*;
import com.intellij.openapi.project.Project;
import com.intellij.psi.*;
import com.intellij.psi.util.PsiTreeUtil;
import org.jetbrains.annotations.NotNull;

import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;
import java.util.concurrent.ConcurrentHashMap;

public class UnittestInspector extends AbstractBaseJavaLocalInspectionTool {

    @Override
    public @NotNull String getGroupDisplayName() {
        return MyBundle.message("inspection.group.name");
    }

    @Override
    public @NotNull String getDisplayName() {
        return MyBundle.message("inspection.display.name");
    }

    @Override
    public @NotNull String getShortName() {
        return "UnittestCaseAnnotationInspection";
    }

    private final Set<String> testAnnotations = new HashSet<>(Arrays.asList(
        "org.junit.Test",
        "org.junit.jupiter.api.Test",
        "Test"
    ));

    private final Set<String> testClassAnnotations = new HashSet<>(Arrays.asList(
        "org.junit.runner.RunWith",
        "org.junit.jupiter.api.TestInstance",
        "org.junit.platform.suite.api.Suite"
    ));

    private final ConcurrentHashMap<String, Boolean> testClassCache = new ConcurrentHashMap<>();

    @Override
    public @NotNull PsiElementVisitor buildVisitor(@NotNull ProblemsHolder holder, boolean isOnTheFly) {
        Project project = holder.getProject();
        return new JavaElementVisitor() {
            @Override
            public void visitMethod(PsiMethod psiMethod) {
                if (!hasTestAnnotation(psiMethod)) {
                    return;
                }

                PsiClass containingClass = psiMethod.getContainingClass();
                if (containingClass == null) {
                    return;
                }
                String qualifiedName = containingClass.getQualifiedName();
                if (qualifiedName == null) {
                    return;
                }
                PsiClass psiClass = PsiTreeUtil.getParentOfType(psiMethod, PsiClass.class);
                if (psiClass == null) {
                    return;
                }
                CaseCheckContext context = CaseCheckContext.create(psiMethod, psiClass);
                new UnittestFileInspectorCommand(holder, project, context).execute();
            }

            private boolean hasTestAnnotation(PsiMethod psiMethod) {
                for (PsiAnnotation annotation : psiMethod.getAnnotations()) {
                if (testAnnotations.contains(annotation.getQualifiedName())) {
                    return true;
                }
            }
                return false;
            }
        };
    }

    @Override
    public @NotNull String getID() {
        return "UnittestCaseAnnotationInspection";
    }

    public void clearCache() {
        testClassCache.clear();
    }
}

File: src/main/kotlin/com/github/jaksonlin/pitestintellij/inspectors/UnittestInspector.kt
package com.github.jaksonlin.pitestintellij.inspectors

import com.github.jaksonlin.pitestintellij.MyBundle
import com.github.jaksonlin.pitestintellij.commands.unittestannotations.UnittestFileInspectorCommand
import com.github.jaksonlin.pitestintellij.context.CaseCheckContext
import com.intellij.codeInspection.*
import com.intellij.psi.*
import com.intellij.psi.util.PsiTreeUtil
import java.util.concurrent.ConcurrentHashMap

class UnittestInspector : AbstractBaseJavaLocalInspectionTool() {
    override fun getGroupDisplayName(): String = MyBundle.message("inspection.group.name")
    override fun getDisplayName(): String = MyBundle.message("inspection.display.name")
    override fun getShortName(): String = "UnittestCaseAnnotationInspection"
    // Cache test annotation qualified names for faster lookup
    private val testAnnotations = setOf(
        "org.junit.Test",
        "org.junit.jupiter.api.Test",
        "Test"
    )

    private val testClassAnnotations = setOf(
        "org.junit.runner.RunWith",
        "org.junit.jupiter.api.TestInstance",
        "org.junit.platform.suite.api.Suite"
    )

    // Cache test class status to avoid repeated checks
    private val testClassCache = ConcurrentHashMap<String, Boolean>()

    override fun buildVisitor(holder: ProblemsHolder, isOnTheFly: Boolean): PsiElementVisitor {
        val project = holder.project
        return object : JavaElementVisitor() {
            override fun visitMethod(psiMethod: PsiMethod) {
                // Quick check for test annotation before doing anything else
                if (!hasTestAnnotation(psiMethod)) {
                    return
                }
                
                // Check containing class using cache
                val containingClass = psiMethod.containingClass ?: return
                val qualifiedName = containingClass.qualifiedName ?: return
                // get the psiclass for the containing class
                val psiClass = PsiTreeUtil.getParentOfType(psiMethod, PsiClass::class.java) ?: return
                val context = CaseCheckContext.create(psiMethod, psiClass)
                UnittestFileInspectorCommand(holder, project, context).execute()
            }

            private fun hasTestAnnotation(psiMethod: PsiMethod): Boolean {
                return psiMethod.annotations.any { annotation ->
                    annotation.qualifiedName in testAnnotations
                }
            }


        }
    }

    override fun getID(): String = "UnittestCaseAnnotationInspection"

    // Clear cache when plugin is unloaded
    fun clearCache() {
        testClassCache.clear()
    }
}